src/frontend/hooks/useMapStyle.ts

```diff
@@ -0,0 +1,154 @@
+import * as React from "react";
+import MapboxGL from "@react-native-mapbox-gl/maps";
+import AsyncStorage from "@react-native-async-storage/async-storage";
+
+import api from "../api";
+import { useNetInfo } from "@react-native-community/netinfo";
+import ky from "ky";
+import { normalizeStyleURL } from "../lib/mapbox";
+import config from "../../config.json";
+import { useExperiments } from "./useExperiments";
+
+/** Key used to store most recent map id in Async Storage */
+const MAP_STYLE_KEY = "@MAPSTYLE";
+
+/** URL used for map style when no custom map and user is online */
+const onlineStyleURL = MapboxGL.StyleURL.Outdoors + "?" + Date.now();
+
+/** URL used for map style when user is not online
+ * generated by [mapeo-offline-map](https://github.com/digidem/mapeo-offline-map) */
+const fallbackStyleURL = "asset://offline-style.json";
+
+type LegacyMapActions =
+  | { action: "setLoading" }
+  | { action: "setCustom"; styleUrl: string }
+  | { action: "setOnline" }
+  | { action: "setFallback" };
+
+type LegacyMapTypes =
+  | { type: "custom" | "online" | "fallback"; styleUrl: string }
+  | { type: "loading"; styleUrl: null };
+
+export type MapTypes =
+  | "loading"
+  | "mapServer"
+  | "custom"
+  | "online"
+  | "fallback";
+
+function legacyMapTypeReducer(
+  state: LegacyMapTypes,
+  action: LegacyMapActions
+): LegacyMapTypes {
+  switch (action.action) {
+    case "setCustom":
+      return { type: "custom", styleUrl: action.styleUrl };
+    case "setOnline":
+      return { type: "online", styleUrl: onlineStyleURL };
+    case "setFallback":
+      return { type: "fallback", styleUrl: fallbackStyleURL };
+    default:
+      return { type: "loading", styleUrl: null };
+  }
+}
+function useLegacyStyle(
+  isOnline: boolean,
+  styleId: string = "default"
+): {
+  styleUrl: string | null;
+  styleType: MapTypes;
+  setStyleId: (id: string) => void;
+} {
+  const [state, dispatch] = React.useReducer(legacyMapTypeReducer, {
+    type: "loading",
+    styleUrl: null,
+  });
+
+  React.useEffect(() => {
+    let didCancel = false;
+
+    // checks if custom maps are available
+    // If not, checks if online maps are available
+    // If not, sets to fallback maps
+    api
+      .getMapStyle(styleId)
+      .then(
+        style => didCancel || dispatch({ action: "setCustom", styleUrl: style })
+      )
+      .catch(() =>
+        ky
+          .get(normalizeStyleURL(onlineStyleURL, config.mapboxAccessToken))
+          .json()
+          .then(() => didCancel || dispatch({ action: "setOnline" }))
+          .catch(() => didCancel || dispatch({ action: "setFallback" }))
+      );
+
+    return () => {
+      didCancel = true;
+    };
+  }, []);
+
+  React.useEffect(() => {
+    if (isOnline && state.type === "fallback") {
+      dispatch({ action: "setOnline" });
+    }
+  }, [isOnline, state.type]);
```

Comment:
Yeah understood, it's best to do that with derived state rather than a useEffect, since bugs are easier to spot (e.g. the current bug if the user is online and goes offlien).

Reply:
But I do see your concern about the infinite loop, but since it is attached to the reducer, we know that when it is `setOnline` the `state.type` is set to `online`